### PR TITLE
Update/add dynamic port binding

### PR DESCRIPTION
**Pull request by xinyuli1204**
the work flow is like this, m3-roca offset is passed as dynamic input and output in building block

1. adding building block like before
```
 auto n = b.add("test_inc_by_offset")(in);
n = b.add("test_inc_by_offset")(n["output"]);
n["output"].bind(out);
```
2. set params field to dynamic ports in node, in test case input param and output param all all points to same address
```
   for(auto &n:b.nodes()){
          for (auto& [pn, port] : n.unbounded_iports()) {
                port.bind(param_buf);
                n.set_iport(port);
          }

          for (auto& [pn, port] : n.unbounded_oports()) {
              port.bind(param_buf);
              n.set_oport(port);
          }

      }
```
3. then  run the builder

the param changed after 2 building block since value in param_buf changed 

| building block index |  input| input_offset |  output_offset | output|
| ------ | ------ | ------ | ------ |  ------ | 
|     No. 1  |     42   | 1    |  1 (input_offset) +1 =2 | 42(input) +1(input_offset) =43
|    No. 2   |    43 | 2|  2 (input_offset)+1 = 3 |  43(input) +2(input_offset) =45

Note: I put compute.root if the input  and output are bound with same address

### CODE REVIEW

File: include/ion/port.h [hunk 1]
```diff
@@ -167,6 +170,7 @@ class Port {
     }
 
     void determine_succ(const NodeID& nid, const std::string& old_pn, const std::string& new_pn);
+    bool is_dnamic_port() const { return impl_->is_dynamic_port; }
```
Comment (richardfixstars) [inline include/ion/port.h:173]:
```suggestion
    bool is_dynamic_port() const { return impl_->is_dynamic_port; }
```
spelling

Comment (xinyuli1204) [reply include/ion/port.h]: fix it

File: include/ion/node.h [hunk 2]
```diff
@@ -121,6 +123,11 @@ class Node {
     Port oport(const std::string& pn);
     std::vector<std::tuple<std::string, Port>> oports() const;
 
+    std::vector<std::tuple<std::string, Port>> dynamic_iports() const;
+    std::vector<std::tuple<std::string, Port>> dynamic_oports() const;
+
+    void  validate_port_address_alignment ()const ;
```
Comment (iitaku) [inline include/ion/node.h:129]: This function looks detecting data hazard. It's better to be more intuitive naming
 https://en.wikipedia.org/wiki/Hazard_(computer_architecture)

File: include/ion/building_block.h [hunk 3]
```diff
@@ -28,6 +28,13 @@ using Input = Halide::GeneratorInput<T>;
 template<typename T>
 using Output = Halide::GeneratorOutput<T>;
 
+template<typename T>
+using DynamicInput = Halide::GeneratorInput<T>;
```
Comment (iitaku) [inline include/ion/building_block.h:32]: The type alias is not necessary for this purpose. You can simply use Input or Output

File: include/ion/node.h [hunk 4]
```diff
@@ -87,6 +87,8 @@ class Node {
 
     void set_iport(const std::string& name, Port port);
 
+    void set_dynamic_port(Port port) ;
```
Comment (iitaku) [inline include/ion/node.h:90]: "dynamic port" is confusing, we already have feature to add input/output dynamically ([ref](https://github.com/fixstars/ion-kit/blob/main/test/test-bb.h#L115)). Can we just say "unbound port"?

File: include/ion/port.h [hunk 5]
```diff
@@ -167,6 +170,7 @@ class Port {
     }
 
     void determine_succ(const NodeID& nid, const std::string& old_pn, const std::string& new_pn);
+    bool is_dynamic() const { return impl_->is_dynamic; }
```
Comment (iitaku) [inline include/ion/port.h:173]: Is it necessary?

File: include/ion/util.h [hunk 6]
```diff
@@ -55,7 +55,7 @@ using NodeID = StringID<node_tag>;
 using GraphID = StringID<graph_tag>;
 using PortID = StringID<port_tag>;
 
-std::string argument_name(const NodeID& node_id, const std::string& name, int32_t index, const GraphID& graph_id);
+std::string argument_name(const PortID & portId, const NodeID& node_id, const std::string& name, int32_t index, const GraphID& graph_id);
```
Comment (iitaku) [inline include/ion/util.h:58]: node_id -> port_id is natural arguments order

File: src/lower.cc [hunk 7]
```diff
@@ -220,6 +220,7 @@ Halide::Pipeline lower(Builder builder, std::vector<Node>& nodes, bool implicit_
     // Constructing Generator object and setting static parameters
     std::unordered_map<NodeID, Halide::Internal::AbstractGeneratorPtr, NodeID::StringIDHash> bbs;
     for (auto n : nodes) {
+        n.validate_port_address_alignment();
```
Comment (iitaku) [inline src/lower.cc:223]: Let's separate loops to keep clear code structure.

File: src/node.cc [hunk 8]
```diff
@@ -138,4 +207,31 @@ std::vector<std::tuple<std::string, Port>> Node::oports() const {
     return oports;
 }
 
+void Node::set_dynamic_port(Port port) {
```
Comment (iitaku) [inline src/node.cc:210]: Can we introduce a generic method to be able to connect ports after constructing the graph?

**Review by iitaku — CHANGES_REQUESTED** [review]

File: src/lower.cc (hunk 7 above)
Comment (xinyuli1204) [reply src/lower.cc:223]:
```
for (auto n : nodes) {
        n.validate_port_address_alignment();
}
```

File: src/node.cc (hunk 8 above)
Comment (xinyuli1204) [reply src/node.cc]:
```
void Node::set_iport(Port port)
void Node::set_oport(Port port)
```

**Comment by iitaku** [comment]
LGTM.

**Review by iitaku — APPROVED** [review]

**Pull request merged**